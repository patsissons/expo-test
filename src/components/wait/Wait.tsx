import React from 'react';
import { View, ActivityIndicator } from 'react-native';
import { Store, AppComponent, withStore } from '../../framework';

export namespace Wait {
  export interface Props {
    isWaiting: (store: Store) => boolean;
    init?: Function;
    header?: any;
    activityIndicator?: any;
  }

  export interface ActivityIndicatorComponentProps {
    animating?: boolean;
    color?: string;
    hidesWhenStopped?: boolean;
    size?: number | 'small' | 'large';
  }

  export interface ComponentProps extends AppComponent.Props<Store>, ActivityIndicatorComponentProps, Props {
  }
}

class WaitComponent extends AppComponent<Wait.ComponentProps, Store> {
  static defaultProps: Partial<Wait.ComponentProps> = {
    size: 'large',
  };

  componentDidMount() {
    if (this.props.init) {
      this.props.init();
    }
  }

  render() {
    if (this.props.isWaiting(this.props.state)) {
      return (
        <View style={ this.S.common.container }>
          { this.R.coerceToText(this.props.header) }
          { this.renderActivityIndicator() }
        </View>
      );
    }

    return this.props.children;
  }

  protected renderActivityIndicator() {
    if (this.props.activityIndicator) {
      return this.props.activityIndicator;
    }

    const { animating, color, hidesWhenStopped, size } = this.props;

    return (
      <ActivityIndicator { ...{ animating, color, hidesWhenStopped, size } } />
    );
  }
}

export const Wait = withStore(
  WaitComponent,
);
