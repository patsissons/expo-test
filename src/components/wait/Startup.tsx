import React from 'react';
import { Store, AppComponent, withStore } from '../../framework';
import { Wait } from './Wait';

export namespace Startup {
  export interface Props {
  }

  export interface ComponentProps extends AppComponent.Props<Store>, Props {
  }
}

class StartupComponent extends AppComponent<Startup.ComponentProps, Store> {
  render() {
    return (
      <Wait init={ this.init.bind(this) } isWaiting={ this.isWaiting.bind(this) }>
        { this.props.children }
      </Wait>
    );
  }

  protected init() {
    this.props.actions.context.loadEnv();
    this.props.actions.context.loadLocale();
  }

  protected isWaiting(store: Store) {
    return [
      store.context == null,
      store.context.env == null,
      store.context.locale == null,
      store.context.localization == null,
    ].some(x => x);
  }
}

export const Startup = withStore(
  StartupComponent,
);
