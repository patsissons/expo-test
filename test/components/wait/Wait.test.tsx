import React from 'react';
import { shallow } from 'enzyme';
import { Text } from 'react-native';
import { createStoreProps } from '../../storeUtils';
import { Wait } from '../../../src/components/wait/Wait';

describe('components', () => {
  describe('Wait', () => {
    it('renders loading mode without crashing', () => {
      // Arrange
      const storeProps = createStoreProps({});
      const component = (<Wait { ...storeProps as any } isWaiting={ () => true } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot('loading');
    });

    it('renders loaded mode without crashing', () => {
      // Arrange
      const storeProps = createStoreProps({});
      const component = (
        <Wait { ...storeProps as any } isWaiting={ () => false }>
          <div>test</div>
        </Wait>
      );

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot('loading');
    });

    it('renders loading mode with custom content', () => {
      // Arrange
      const storeProps = createStoreProps({ context: {} });
      const content = (<Text testID='custom'>test</Text>);
      const component = (<Wait { ...storeProps as any } isWaiting={ () => true } activityIndicator={ content } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result
        .find({ testID: 'custom' })
        .children()
        .text(),
      ).toBe('test');
    });

    it('renders loading mode with custom content', () => {
      const storeProps = createStoreProps({});
      const init = jest.fn();
      const component = (<Wait { ...storeProps as any } isWaiting={ () => true } init={ init } />);

      // Act
      shallow(component).dive();

      // Assert
      expect(init).toHaveBeenCalled();
    });
  });
});
