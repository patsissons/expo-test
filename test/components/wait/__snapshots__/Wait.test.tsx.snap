// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Wait renders loaded mode without crashing: loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WaitComponent
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={
      Object {
        "env": Object {
          "RELEASE": "local",
        },
        "locale": "en",
        "localization": Object {
          "CurrentStateIs": "Current state is",
          "Increment": "Increment",
          "IncrementCount": "Increment Count",
          "Login": "Login",
          "LoginScreen": "Login Screen",
          "MainScreen": "Main Screen",
        },
      }
    }
    isWaiting={[Function]}
    size="large"
    state={
      Object {
        "context": Object {
          "env": Object {
            "RELEASE": "local",
          },
          "locale": "en",
          "localization": Object {
            "CurrentStateIs": "Current state is",
            "Increment": "Increment",
            "IncrementCount": "Increment Count",
            "Login": "Login",
            "LoginScreen": "Login Screen",
            "MainScreen": "Main Screen",
          },
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <div>
      test
    </div>
  </WaitComponent>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "test",
    },
    "ref": null,
    "rendered": "test",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "test",
      },
      "ref": null,
      "rendered": "test",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`components Wait renders loading mode without crashing: loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WaitComponent
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={
      Object {
        "env": Object {
          "RELEASE": "local",
        },
        "locale": "en",
        "localization": Object {
          "CurrentStateIs": "Current state is",
          "Increment": "Increment",
          "IncrementCount": "Increment Count",
          "Login": "Login",
          "LoginScreen": "Login Screen",
          "MainScreen": "Main Screen",
        },
      }
    }
    isWaiting={[Function]}
    size="large"
    state={
      Object {
        "context": Object {
          "env": Object {
            "RELEASE": "local",
          },
          "locale": "en",
          "localization": Object {
            "CurrentStateIs": "Current state is",
            "Increment": "Increment",
            "IncrementCount": "Increment Count",
            "Login": "Login",
            "LoginScreen": "Login Screen",
            "MainScreen": "Main Screen",
          },
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        false,
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="large"
        />,
      ],
      "style": Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animating": true,
          "color": "#999999",
          "hidesWhenStopped": true,
          "size": "large",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          false,
          <ActivityIndicator
            animating={true}
            color="#999999"
            hidesWhenStopped={true}
            size="large"
          />,
        ],
        "style": Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animating": true,
            "color": "#999999",
            "hidesWhenStopped": true,
            "size": "large",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
